configuration/vpn.md

```diff
@@ -60,73 +60,128 @@ Using a ProxyVM to set up a VPN client gives you the ability to:
+    iptables -F OUTPUT
+
+	#    Add the `qvpn` group to system, if it doesn't already exist
+    if ! grep -q "^qvpn:" /etc/group ; then
+        groupadd -rf qvpn
+        sync
+    fi
+    sleep 2s
+
+	#    Allow traffic from the `qvpn` group to the uplink interface (eth0);
+	#    Our openvpn will run as group `qvpn`.
+    iptables -A OUTPUT -p all -o eth0 -m owner --gid-owner qvpn \
+    -m state --state NEW,ESTABLISHED -j ACCEPT
+
+	#	Allow queries to DNS server:
+    iptables -A OUTPUT -p udp -o eth0 --dport 53 -m state --state NEW,ESTABLISHED -j ACCEPT
+    iptables -A OUTPUT -p tcp -o eth0 --dport 53 -m state --state NEW,ESTABLISHED -j ACCEPT
+
+	#	Allow internal system connections:
+    iptables -I OUTPUT -o lo -j ACCEPT
+
+	#	Block forwarding of connections through upstream network device
+	#	(in case the vpn tunnel breaks):
+    iptables -I FORWARD -o eth0 -j DROP  
+    iptables -I FORWARD -i eth0 -j DROP
+    ```
     Now save `/rw/config/qubes-firewall-user-script` and make it executable:  
     `sudo chmod +x /rw/config/qubes-firewall-user-script`
-    
-4.  Setup the vpn's autostart:  
-   Edit to `/rw/config/rc.local`, make it executable  (`sudo chmod +x /rw/config/rc.local`) and add:  
 
-        ln -s /home/user/vpn.cfg /etc/openvpn/vpn.conf;  
-        systemctl --no-block start openvpn@vpn.service;
+4.  Create the DNS-handling script.
+    Use `sudo nano /rw/config/openvpn/qubes-vpn-handler.sh` to edit and add:
+    ```
+    #!/bin/bash
+    set -e
+
+# Pop-up notification variables
+SPID=$(pgrep -U user -f dconf-service)
+dbus=$(grep -z DBUS_SESSION_BUS_ADDRESS /proc/$SPID/environ|cut -d= -f2-)
+export DBUS_SESSION_BUS_ADDRESS=$dbus
+
+case "$1" in
+
+up)
+	# To override DHCP DNS, assign static DNS addresses with 'setenv vpn_dns' in openvpn config;
+	# Format is 'X.X.X.X  Y.Y.Y.Y [...]' with quotes.
+	if [[ -z $vpn_dns ]] ; then
+		# Parses DHCP options from openvpn to set DNS address translation:
+		for optionname in ${!foreign_option_*} ; do
+			option="${!optionname}"
+			unset fops; fops=($option)
+			if [ ${fops[1]} == "DNS" ] ; then vpn_dns="$vpn_dns ${fops[2]}" ; fi
+		done
+	fi
+
+	iptables -t nat -F PR-QBS
+	if [[ -n $vpn_dns ]] ; then
+		# Set DNS address translation in firewall:
+		for addr in $vpn_dns; do
+			iptables -t nat -A PR-QBS -i vif+ -p udp --dport 53 -j DNAT --to $addr
+			iptables -t nat -A PR-QBS -i vif+ -p tcp --dport 53 -j DNAT --to $addr
+		done
+		su -c 'notify-send "$(hostname): LINK IS UP." --icon=network-idle' user
+	else
+		su -c 'notify-send "$(hostname): LINK UP, NO DNS!" --icon=dialog-error' user
+	fi
+
+	;;
+down)
+	su -c 'notify-send "$(hostname): LINK IS DOWN !" --icon=dialog-error' user
+	;;
+esac
+```
+
+    Now save the script and make it executable:  
+    `sudo chmod +x /rw/config/openvpn/qubes-vpn-handler.sh`
+    
+5.  Setup the VPN's autostart:  
+    Use `sudo nano /rw/config/rc.local` to edit and add:  
+    ```
+    #!/bin/bash
+    groupadd -rf qvpn ; sleep 2s
```

Comment:
Independently.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?